File: backend/internal/entity/custom_page.go
package entity

import (
	"encoding/json"
	"time"
)

// ComponentType represents the type of a page component
type ComponentType string

const (
	ComponentTypeIframe     ComponentType = "iframe"
	ComponentTypeText       ComponentType = "text"
	ComponentTypeMarkdown   ComponentType = "markdown"
	ComponentTypeHTML       ComponentType = "html"
	ComponentTypeEntityList ComponentType = "entity_list"
	ComponentTypeLinkGroup  ComponentType = "link_group"
	ComponentTypeStats      ComponentType = "stats"
)

// ComponentWidth represents the width of a component in the layout
type ComponentWidth string

const (
	ComponentWidthFull     ComponentWidth = "full"
	ComponentWidthHalf     ComponentWidth = "1/2"
	ComponentWidthThird    ComponentWidth = "1/3"
	ComponentWidthTwoThird ComponentWidth = "2/3"
)

// PageComponent represents a single component on a custom page
type PageComponent struct {
	ID     string          `json:"id"`
	Type   ComponentType   `json:"type"`
	Title  string          `json:"title,omitempty"`
	Width  ComponentWidth  `json:"width"`
	Order  int             `json:"order"`
	Config json.RawMessage `json:"config"`
}

// IframeConfig holds configuration for an iframe component
type IframeConfig struct {
	URL     string `json:"url"`
	Height  int    `json:"height,omitempty"`
	Sandbox string `json:"sandbox,omitempty"`
}

// TextConfig holds configuration for a text/markdown component
type TextConfig struct {
	Content string `json:"content"`
}

// HTMLConfig holds configuration for a raw HTML component
type HTMLConfig struct {
	Content string `json:"content"`
}

// EntityListConfig holds configuration for an entity list component
type EntityListConfig struct {
	Entity   string                 `json:"entity"`
	Filters  map[string]interface{} `json:"filters,omitempty"`
	Columns  []string               `json:"columns,omitempty"`
	PageSize int                    `json:"pageSize,omitempty"`
	SortBy   string                 `json:"sortBy,omitempty"`
	SortDir  string                 `json:"sortDir,omitempty"`
}

// LinkItem represents a single link in a link group
type LinkItem struct {
	Label       string `json:"label"`
	Href        string `json:"href"`
	Icon        string `json:"icon,omitempty"`
	Description string `json:"description,omitempty"`
	External    bool   `json:"external,omitempty"`
}

// LinkGroupConfig holds configuration for a link group component
type LinkGroupConfig struct {
	Links []LinkItem `json:"links"`
}

// StatItem represents a single stat in a stats component
type StatItem struct {
	Label string `json:"label"`
	Value string `json:"value"` // Can contain templates like {{count:contacts}}
	Icon  string `json:"icon,omitempty"`
	Color string `json:"color,omitempty"`
}

// StatsConfig holds configuration for a stats component
type StatsConfig struct {
	Items []StatItem `json:"items"`
}

// CustomPage represents a custom page in the database
type CustomPage struct {
	ID          string          `json:"id" db:"id"`
	OrgID       string          `json:"orgId" db:"org_id"`
	Slug        string          `json:"slug" db:"slug"`
	Title       string          `json:"title" db:"title"`
	Description string          `json:"description,omitempty" db:"description"`
	Icon        string          `json:"icon" db:"icon"`
	IsEnabled   bool            `json:"isEnabled" db:"is_enabled"`
	IsPublic    bool            `json:"isPublic" db:"is_public"`
	Layout      string          `json:"layout" db:"layout"`
	Components  []PageComponent `json:"components" db:"-"`
	SortOrder   int             `json:"sortOrder" db:"sort_order"`
	CreatedAt   time.Time       `json:"createdAt" db:"created_at"`
	ModifiedAt  time.Time       `json:"modifiedAt" db:"modified_at"`
	CreatedBy   *string         `json:"createdBy,omitempty" db:"created_by"`
	ModifiedBy  *string         `json:"modifiedBy,omitempty" db:"modified_by"`
}

// CustomPageCreateInput for creating a new custom page
type CustomPageCreateInput struct {
	Slug        string          `json:"slug" validate:"required"`
	Title       string          `json:"title" validate:"required"`
	Description string          `json:"description,omitempty"`
	Icon        string          `json:"icon,omitempty"`
	IsEnabled   bool            `json:"isEnabled"`
	IsPublic    bool            `json:"isPublic"`
	Layout      string          `json:"layout,omitempty"`
	Components  []PageComponent `json:"components,omitempty"`
	SortOrder   int             `json:"sortOrder,omitempty"`
}

// CustomPageUpdateInput for updating an existing custom page
type CustomPageUpdateInput struct {
	Slug        *string          `json:"slug,omitempty"`
	Title       *string          `json:"title,omitempty"`
	Description *string          `json:"description,omitempty"`
	Icon        *string          `json:"icon,omitempty"`
	IsEnabled   *bool            `json:"isEnabled,omitempty"`
	IsPublic    *bool            `json:"isPublic,omitempty"`
	Layout      *string          `json:"layout,omitempty"`
	Components  *[]PageComponent `json:"components,omitempty"`
	SortOrder   *int             `json:"sortOrder,omitempty"`
}

// CustomPageListItem is a lightweight version for listing pages
type CustomPageListItem struct {
	ID          string    `json:"id"`
	Slug        string    `json:"slug"`
	Title       string    `json:"title"`
	Description string    `json:"description,omitempty"`
	Icon        string    `json:"icon"`
	IsEnabled   bool      `json:"isEnabled"`
	IsPublic    bool      `json:"isPublic"`
	SortOrder   int       `json:"sortOrder"`
	ModifiedAt  time.Time `json:"modifiedAt"`
}

File: frontend/src/lib/types/custom-page.ts
// Component types supported on custom pages
export type ComponentType = 'iframe' | 'text' | 'markdown' | 'html' | 'entity_list' | 'link_group' | 'stats';

// Component width in the grid layout
export type ComponentWidth = 'full' | '1/2' | '1/3' | '2/3';

// Base component interface
export interface PageComponent {
	id: string;
	type: ComponentType;
	title?: string;
	width: ComponentWidth;
	order: number;
	config: IframeConfig | TextConfig | HTMLConfig | EntityListConfig | LinkGroupConfig | StatsConfig;
}

// Component-specific configurations
export interface IframeConfig {
	url: string;
	height?: number;
	sandbox?: string;
}

export interface TextConfig {
	content: string;
}

export interface HTMLConfig {
	content: string;
}

export interface EntityListConfig {
	entity: string;
	filters?: Record<string, unknown>;
	columns?: string[];
	pageSize?: number;
	sortBy?: string;
	sortDir?: 'asc' | 'desc';
}

export interface LinkItem {
	label: string;
	href: string;
	icon?: string;
	description?: string;
	external?: boolean;
}

export interface LinkGroupConfig {
	links: LinkItem[];
}

export interface StatItem {
	label: string;
	value: string; // Can contain templates like {{count:contacts}}
	icon?: string;
	color?: string;
}

export interface StatsConfig {
	items: StatItem[];
}

// Custom page interfaces
export interface CustomPage {
	id: string;
	orgId: string;
	slug: string;
	title: string;
	description?: string;
	icon: string;
	isEnabled: boolean;
	isPublic: boolean;
	layout: 'single' | 'grid';
	components: PageComponent[];
	sortOrder: number;
	createdAt: string;
	modifiedAt: string;
	createdBy?: string;
	modifiedBy?: string;
}

export interface CustomPageListItem {
	id: string;
	slug: string;
	title: string;
	description?: string;
	icon: string;
	isEnabled: boolean;
	isPublic: boolean;
	sortOrder: number;
	modifiedAt: string;
}

export interface CustomPageCreateInput {
	slug: string;
	title: string;
	description?: string;
	icon?: string;
	isEnabled?: boolean;
	isPublic?: boolean;
	layout?: 'single' | 'grid';
	components?: PageComponent[];
	sortOrder?: number;
}

export interface CustomPageUpdateInput {
	slug?: string;
	title?: string;
	description?: string;
	icon?: string;
	isEnabled?: boolean;
	isPublic?: boolean;
	layout?: 'single' | 'grid';
	components?: PageComponent[];
	sortOrder?: number;
}

// Helper constants
export const COMPONENT_TYPES: { value: ComponentType; label: string; description: string; icon: string }[] = [
	{ value: 'iframe', label: 'Iframe', description: 'Embed an external website or app', icon: 'globe' },
	{ value: 'text', label: 'Text', description: 'Plain text content', icon: 'align-left' },
	{ value: 'markdown', label: 'Markdown', description: 'Rich text with markdown formatting', icon: 'file-text' },
	{ value: 'html', label: 'HTML', description: 'Custom HTML content', icon: 'code' },
	{ value: 'entity_list', label: 'Entity List', description: 'Display records from an entity', icon: 'list' },
	{ value: 'link_group', label: 'Quick Links', description: 'Group of shortcut links', icon: 'link' },
	{ value: 'stats', label: 'Stats', description: 'Display key metrics', icon: 'bar-chart-2' }
];

export const COMPONENT_WIDTHS: { value: ComponentWidth; label: string }[] = [
	{ value: 'full', label: 'Full Width' },
	{ value: '1/2', label: 'Half (1/2)' },
	{ value: '1/3', label: 'One Third (1/3)' },
	{ value: '2/3', label: 'Two Thirds (2/3)' }
];

export const PAGE_ICONS: string[] = [
	'file', 'folder', 'home', 'dashboard', 'chart', 'bar-chart', 'pie-chart',
	'users', 'user', 'settings', 'star', 'heart', 'bookmark', 'flag',
	'calendar', 'clock', 'mail', 'phone', 'globe', 'map', 'link',
	'document', 'clipboard', 'archive', 'briefcase', 'dollar-sign', 'percent'
];
